From: Santiago Vila <sanvila@debian.org>
Subject: We can do fine without the build date

--- a/src/tla/tla/tla.c
+++ b/src/tla/tla/tla.c
@@ -61,9 +61,8 @@
     {
       safe_printfmt (1, cfg__std__package "\n");
 #ifdef TLA_REVISION
-      safe_printfmt(1, "%s, ", TLA_REVISION);
+      safe_printfmt(1, "%s\n", TLA_REVISION);
 #endif
-      safe_printfmt (1, "built on %s\n", __DATE__);
       safe_printfmt (1, "\n");
       safe_printfmt (1, "Copyright 2003, 2004, 2005, 2006 Free Software Foundation, Inc. and contributors\n");
       safe_printfmt (1, "Written by Thomas Lord, et al. See the AUTHORS file for a list.\n");
